models/spectral_clustering.py

```diff
@@ -0,0 +1,279 @@
+import shapes
+import math
+import functools
+import operator
+import typing
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import numpy as np
+
+def cosine_kernel(E):
+	E = F.normalize(E, dim = 1)
+	return E @ E.t()
+
+
+def normalized_symmetric_laplacian(W):
+	# https://en.wikipedia.org/wiki/Laplacian_matrix#Symmetric_normalized_Laplacian
+	D_sqrt = W.sum(dim = -1).sqrt()
+	return torch.eye(W.shape[-1], device = W.device, dtype = W.dtype) - W / D_sqrt.unsqueeze(-1) / D_sqrt.unsqueeze(-2)
+
+
+class SpectralClusteringDiarizationModel:
+	def __init__(self, vad, vad_sensitivity: float, weights_path: str, sample_rate: int):
+		self.vad = vad
+		self.vad_sensitivity = vad_sensitivity
+		self.sample_rate = sample_rate
+		self.embed_model = SincTDNN(sample_rate = sample_rate, padding_same = True, tdnn = dict(kernel_size_pool = 501, stride_pool = 101))
+		diag = self.embed_model.load_state_dict(torch.load(weights_path, map_location='cpu'), strict=False)
```

Comment:
please extract pyannote checkpoint loading to a separate method `load_pyannote_checkpoint` + don't call it from constructor, let the callsite call it

Reply:
Motivation? Object of class PyannoteDiarizationModel cant work properly without loading.